src/HttpServer.php

```diff
@@ -60,50 +64,102 @@ public function addMiddleware($middleware)
+        $listener =  function ($data) use (&$connection, $that, &$headerBuffer, &$listener) {
+            $headerBuffer .= $data;
+            if (strpos($headerBuffer, "\r\n\r\n") !== false) {
+                $connection->removeListener('data', $listener);
+                // header is completed
+                $fullHeader = substr($headerBuffer, 0, strpos($headerBuffer, "\r\n\r\n") + 4);
 
-        $headerStream = new ReadableStream();
-        $headerDecoder = new HeaderDecoder($headerStream);
+                try {
+                    $request = RingCentral\Psr7\parse_request($fullHeader);
+                } catch (\Exception $ex) {
+                    $that->sendResponse(new Response(400), $connection);
+                    return;
+                }
+                // remove header from $data, only body is left
+                $data = (string)substr($data, strlen($fullHeader));
 
-        $connection->on('data', function ($data) use ($connection, &$headerCompleted, &$bodyBuffer, $that, &$chunkedDecoder, &$headerDecoder, $chunkStream, $headerStream) {
-            if (!$headerCompleted) {
-                $headerDecoder->on('data', function ($header) use (&$request, &$data, &$headerCompleted) {
-                    $request = RingCentral\Psr7\parse_request($header);
-                    $data = substr($data, strlen($header));
-                    $headerCompleted = true;
-                });
-                $headerStream->emit('data', array($data));
-            }
+                $that->handleBody($request, $connection);
 
-            if (isset($request)) {
-                if ($that->isChunkedEncodingActive($request)) {
-                    $chunkedDecoder->on('data', function ($chunk) use (&$bodyBuffer, &$request, $that, $connection) {
-                        $bodyBuffer .= $chunk;
-                        if (strlen($chunk) == 0) {
-                            $that->sendBody($bodyBuffer, $connection, $request);
-                        }
-                    });
-                    $chunkStream->emit('data', array($data));
-                } else {
-                    $bodyBuffer .= $data;
-                    $contentLengthArray = $request->getHeader('Content-Length');
-
-                    if (!empty($contentLengthArray) && strlen($bodyBuffer) == $contentLengthArray[0]) {
-                        $that->sendBody($bodyBuffer, $connection, $request);
-                    } else if (empty($contentLengthArray) || $contentLengthArray[0] == 0) {
-                        $that->handleRequest($connection, $request);
-                    }
+                if ($data !== '') {
+                    $connection->emit('data', array($data));
                 }
             }
+        };
+
+        $connection->on('data', $listener);
+    }
+
+    /** @internal */
+    public function handleBody(RequestInterface $request, ConnectionInterface $connection)
+    {
+        $bodyDataStream = new ReadableStream();
```

Comment:
Does not support backpressure etc. and will likely be removed in future versions (see older comment)